examples/step-27/step-27.cc

```diff
@@ -474,103 +418,70 @@ namespace Step27
-  // there, i.e. the square domain with the square hole in the middle. The
-  // meaning of the different parts of this function are explained in the
-  // documentation of step-14:
-  template <>
-  void LaplaceProblem<2>::create_coarse_grid()
+  // The following function is used when creating the initial grid. The grid we
+  // would like to create is actually similar to the one from step-14, i.e., the
+  // square domain with the square hole in the middle. It can be generated by
+  // excatly the same function. However, since its implementation is only a
+  // specialization of the 2d case, we will present a different way of creating
+  // this domain which is dimension independent.
+  //
+  // We first create a hypercube triangulation with enough cells so that it
+  // already holds our desired domain. We then continue to remove those cells in
+  // the center of the domain by identifying them with the physical coordinates
+  // of the vertices on each cell. In the end, we refine the so created grid
+  // globally as usual.
+  template <int dim>
+  void LaplaceProblem<dim>::create_coarse_grid()
   {
-    const unsigned int dim = 2;
-
-    const std::vector<Point<2>> vertices = {
-      {-1.0, -1.0}, {-0.5, -1.0}, {+0.0, -1.0}, {+0.5, -1.0}, {+1.0, -1.0}, //
-      {-1.0, -0.5}, {-0.5, -0.5}, {+0.0, -0.5}, {+0.5, -0.5}, {+1.0, -0.5}, //
-      {-1.0, +0.0}, {-0.5, +0.0}, {+0.5, +0.0}, {+1.0, +0.0},               //
-      {-1.0, +0.5}, {-0.5, +0.5}, {+0.0, +0.5}, {+0.5, +0.5}, {+1.0, +0.5}, //
-      {-1.0, +1.0}, {-0.5, +1.0}, {+0.0, +1.0}, {+0.5, +1.0}, {+1.0, +1.0}};
-
-    const std::vector<std::array<int, GeometryInfo<dim>::vertices_per_cell>>
-      cell_vertices = {{{0, 1, 5, 6}},
-                       {{1, 2, 6, 7}},
-                       {{2, 3, 7, 8}},
-                       {{3, 4, 8, 9}},
-                       {{5, 6, 10, 11}},
-                       {{8, 9, 12, 13}},
-                       {{10, 11, 14, 15}},
-                       {{12, 13, 17, 18}},
-                       {{14, 15, 19, 20}},
-                       {{15, 16, 20, 21}},
-                       {{16, 17, 21, 22}},
-                       {{17, 18, 22, 23}}};
-
-    const unsigned int n_cells = cell_vertices.size();
-
-    std::vector<CellData<dim>> cells(n_cells, CellData<dim>());
-    for (unsigned int i = 0; i < n_cells; ++i)
-      {
-        for (unsigned int j = 0; j < cell_vertices[i].size(); ++j)
-          cells[i].vertices[j] = cell_vertices[i][j];
-        cells[i].material_id = 0;
-      }
+    Triangulation<dim> cube;
+    GridGenerator::subdivided_hyper_cube(cube, 4, -1., 1.);
+
+    std::set<typename Triangulation<dim>::active_cell_iterator> cells_to_remove;
+    for (const auto &cell : cube.active_cell_iterators())
+      for (unsigned int v = 0; v < GeometryInfo<dim>::vertices_per_cell; ++v)
+        for (unsigned int d = 0; d < dim; ++d)
+          if (std::abs(cell->vertex(v)[d]) < .5)
+            cells_to_remove.insert(cell);
```

Comment:
We might use `std::copy_if` here instead, but I do not know if this would confuse users who are new to C++.

Reply:
I think this is ok.